### PR TITLE
Misc fixes

### LINKED ISSUE #267: matrix slicing broken
```
When trying to manipulate matrices using mpmath, I'm confronted with an obstacle.  Using scipy/numpy/etc. one can call (and thus modify/alter/etc.) a "section" of a matrix, i.e., slicing the matrix.  However, with mpmath.matrix I'm unable to do so.  Is there an acceptable workaround to this. What steps will reproduce the problem? >>> from sympy import Matrix as sm_matrix
>>> from scipy import matrix as sp_matrix
>>> from mpmath import matrix as mp_matrix
>>> 
>>> my = sm_matrix([[1,2,3],[4,5,6],[7,8,9]])
>>> mc = sp_matrix([[1,2,3],[4,5,6],[7,8,9]])
>>> mp = mp_matrix([[1,2,3],[4,5,6],[7,8,9]])
>>> 
>>> my[0:2,0:2]
[1, 2]
[4, 5]
>>> mc[0:2,0:2]
matrix([[1, 2],
        [4, 5]])
>>> mp[0:2,0:2]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/pymodules/python2.7/mpmath/matrices/matrices.py", line 412, in __getitem__
    if key in self.__data:
TypeError: unhashable type What is the expected output? What do you see instead? I would have expected a similar response to the prior two commands... a submatrix consisting of the elements [[1,2],[4,5]].  Am I wrong in assuming? What version of the product are you using? On what operating system? mpmath.__version__ = 0.14 & 0.15 (two different machines)
Both operating Ubuntu Linux.

Thanks for any assistance you can give.

~~archery~~
```

Original issue for #267: http://code.google.com/p/mpmath/issues/detail?id=227
Original author: https://code.google.com/u/archeryguru2000/

### LINKED ISSUE #274: segmentation fault when calling the mpmath.mp.fraction(1,100).func(1000,None)
```
What steps will reproduce the problem? 1. make sure gmpy is installed
2. import mpmath
3. mpmath.mp.fraction(1,100).func(1000, 0xdead ) What is the expected output? What do you see instead? Some kind of exception. The python backend produces a KeyError What version of the product are you using? On what operating system? Debian testing distribution
    python-mpmath = 0.17-1
    python-gmpy = 1.15-1 Please provide any additional information below. When the type of the last parameter is not correct, (say None, or an integer) A segmentation fault happens, when using gmpy.
```

Original issue for #274: http://code.google.com/p/mpmath/issues/detail?id=234
Original author: https://code.google.com/u/116125508345498896876/
Original owner: https://code.google.com/u/casevh/

### LINKED ISSUE #283: Equality testing for matrices assumes that entities tested are, in fact, matrices
```
What steps will reproduce the problem? import mpmath as MP
BigMat=None # initialize a symbol that might or might not be used

# Work that may or may not create BigMat depending on circumstances
# but which, in a particular instance, does happen to execute...
BigMat=MP.matrix([1,2,3])
# ... and then later, to see if the program took that path one tries..

if BigMat==None:
   blah..
else:
   blah blah... 

What is the expected output? 
The expected result would be a False or True branch in the "if" statement
depending on whether BigMat had or had not been created by the particular workflow.

What do you see instead?
    669     def __eq__(self, other):
--> 670         return self.__rows == other.__rows and self.__cols == other.__cols \
    671                and self.__data == other.__data
    672 

AttributeError: 'NoneType' object has no attribute '_matrix__rows' What version of the product are you using? On what operating system? 0.17 on Windows 7 (64-bit) Please provide any additional information below. Entities can fail to be equal for lots of reasons; a solution would be to preface tests with guard statements generating appropriate returns when apples and oranges are compared.
```

Original issue for #283: http://code.google.com/p/mpmath/issues/detail?id=243
Original author: https://code.google.com/u/115096564883933431558/

### LINKED ISSUE #377: from_str: wrong parsing with underscores after decimal
This works:
````
>>> mpmath.libmp.from_str('1_234.567891', 80)
(0, mpz(91095031707585398740483), -66, 77)
>>> mpmath.libmp.to_str(_, 24)
'1234.567891'
````

But underscores placed *after* the decimal point shift the decimal point incorrectly:
````
>>> mpmath.libmp.from_str('1_234.567_891', 80)
(0, mpz(1166016405857093103878183), -73, 80)
>>> mpmath.libmp.to_str(_, 24)
'123.4567891'

>>> mpmath.libmp.from_str('1_234.567_8_9_1', 80)
(0, mpz(746250499748539586482037), -79, 80)
>>> mpmath.libmp.to_str(_, 24)
'1.234567891'
````

The docs say "The literal syntax accepted is the same as for Python floats".  But this is not so:
````
>>> 1_234.567_8_9_1
1234.567891
````

### LINKED ISSUE #428: /doc and README.rst missing on pypi
The tar-ball on github seems to have the doc and README.rst file, but the one from pypi is missing these.  (Trying to package the lastest version for opensuse and the standard is to get the tar-balls from pypi). Would be great to include these again in the next version.

Thanks

### LINKED ISSUE #539: mpmath.mpf fails converting numpy.int64
Hi! 

I understand there was a very similar issue #382 that is closed, but it appears I'm getting the same issue again.

```
>>from mpmath import mpf
>>import numpy

>>my_mpf = mpf(numpy.int64(1))

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "...site-packages/mpmath/ctx_mp_python.py", line 77, in __new__
    v._mpf_ = mpf_pos(cls.mpf_convert_arg(val, prec, rounding), prec, rounding)
  File "...site-packages/mpmath/ctx_mp_python.py", line 96, in mpf_convert_arg
    raise TypeError("cannot create mpf from " + repr(x))
TypeError: cannot create mpf from 1

```

To my understanding, I'm using the latest version of mpmath 
```
>> import mpmath
>> mpmath.__version__
1.1.0
```
python --version 
3.7.6

Thanks!

### PR DESCRIPTION
**Pull request by skirpichev**
* fix #283
* add regression tests, closes #267, closes #539, closes #274
* run CI tests with numpy
* allow `mpf('.00')` like float constructor does, closes #377 
* add docs/ to sdist, closes #428

### CODE REVIEW

File: mpmath/matrices/matrices.py [hunk 1]
```diff
@@ -676,8 +676,11 @@ def __rsub__(self, other):
         return -self + other
 
     def __eq__(self, other):
-        return self.__rows == other.__rows and self.__cols == other.__cols \
-               and self.__data == other.__data
+        try:
+            return (self.__rows == other.__rows and self.__cols == other.__cols
+                    and self.__data == other.__data)
+        except AttributeError:
+            return NotImplemented
```
Comment (cbm755) [inline mpmath/matrices/matrices.py:683]: I'd not seen this before, but I skimmed https://stackoverflow.com/questions/40780004/returning-notimplemented-from-eq

I also tested this, and confirm that the `A9 == None` test below.

**Review by cbm755 — APPROVED** [review]
This looks good to me!

The `build` packaging dependency probably should've been a different PR: I've never used that before, but I looked it up and it seems reasonable to me.

https://pypi.org/project/build/  <-- "A simple, correct Python build frontend."

File: MANIFEST.in [hunk 2]
```diff
@@ -0,0 +1,3 @@
+recursive-include docs *
+include CHANGES
+exclude MANIFEST.in
```
[the remark this answers is not in the document]
Comment (skirpichev) [reply MANIFEST.in:3]: Not sure, this was blindly copied from the SymPy's file, where this line was added in sympy/sympy#24363 (without comments).  Lets remove this.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -53,7 +53,7 @@ jobs:
         make -C build/sphinx/latex all-pdf
     - name: Make packages
       if: matrix.python-version == 3.11
-      run: python setup.py sdist bdist_wheel
+      run: python -m build
```
[the remark this answers is not in the document]
Comment (skirpichev) [reply .github/workflows/test.yml:56]: Yep, but you were right: probably this should go to a separate pr.

**Review by cbm755 — APPROVED** [review]

**Comment by cbm755** [comment]
LGTM!
